Feature/mesher and planner plugins

woodaaron opened the pull request:
Refactoring of Godel workflow to incorporate meshing and path-planning plugins.
Decouples path-planning from visualization of plans
Decouples edge, scan, and blend planning so they may be exercised independently
Moves mesh_importer and profilometer_scan to path_planning_plugins
Removes some configuration options from the "admin" GUI related to blend configuration
Added configuration yaml files to control scan/blend parameters

Commit:
Added default meshing plugin and integrated into surface_detection
Added stub for path planning plugin

Commit:
Decoupled visualization from path plan generation
Created new message for path planning without visualization
Changed surface detection client to request planning w/o visualization

Commit:
Replaced blend configuration options with path planning options in GUI
Added PathPlanning service and PathPlanningParameters message
Added PathPlanningParameters message to SurfaceBlendingParameters service
Replaced Blend/ScanPlanparameters in ProcessPlanning service

Commit:
Decoupled visualization from path planning
Renamed remove_pevious_process_plan function to clear_visualizations
Added publisher for scan poses

Commit:
Refactored generateProcessPath so that blend, scan, and edge have independent path generations
Removed requestScan/Blend/EdgePath functions

Commit:
Simplified base interface and separated implementations to Scan and Blend types

Commit:
Moved meshing/path-planning plugins from godel_plugins to their own packages

Commit:
Moved mesh_importer and "profilimeter_scan" (now profilometer_scan) to path_planning_plugins

Commit:
Implemented path-planning plugin
Added rosparam tag to load parameter config yamls on launch
Added process planning parameters config yaml

Jmeyer1292 commented on godel_plugins/CMakeLists.txt:
```diff
@@ -13,12 +13,13 @@ find_package(catkin REQUIRED COMPONENTS
 find_package(Qt5Widgets REQUIRED)
 find_package(Qt5Core REQUIRED)
 find_package(Qt5Gui REQUIRED)
+find_package(PCL REQUIRED)
```
Does this package need all the PCL dependencies?

woodaaron replied:
Nope, that was a holdover from a previous implementation. I'll remove it.

Jmeyer1292 commented on godel_process_path_generation/src/process_path.cpp:
```diff
@@ -55,4 +56,20 @@ visualization_msgs::Marker ProcessPath::asMarker() const
+    geometry_msgs::Pose pose;
+    pose.position.x = pt.pose().translation()(0);
+    pose.position.y = pt.pose().translation()(1);
+    pose.position.z = pt.pose().translation()(2);
```
This makes me oh-so-slightly uncomfortable because it does not set any of the orientation fields of the `pose` object and thus it has an invalid rotation matrix. What should the pose be of these objects? If it's just identity, then set `pose.orientation.w = 1` or something like that.

Jmeyer1292 commented on godel_robots/abb/godel_irb2400/godel_irb2400_support/launch/irb2400_blending.launch:
```diff
@@ -76,5 +76,8 @@
   <node pkg="tf" type="static_transform_publisher" name="ensenso_static_tf" args="0 0 0 0 1.57 0 kinect2_move_frame camera_optical_frame 100" />
+  <rosparam command="load" file="$(find meshing_plugins)/config/meshing.yaml" />
+  <rosparam command="load" file="$(find path_planning_plugins)/config/path_planning.yaml" />
+  <rosparam command="load" file="$(find godel_process_planning)/config/process_planning.yaml" />
```
I initially imagined that configuration files would live in the `workcell` specific packages like `godel_irb2400_support`. Can you elaborate on the decision to put them in packages that implement certain functionality?

woodaaron replied:
The idea was primarily that individual implementations of specific plugins might require different parameters. Keeping the configuration file in the folder where the plugin lives encapsulates the plugin and makes it easier to maintain.

Wrt process planning, it was just a matter of convenience. I'm fine with moving it around.

Jmeyer1292 commented on godel_surface_detection/src/detection/surface_detection.cpp:
```diff
@@ -15,10 +15,12 @@
 #include <godel_param_helpers/godel_param_helpers.h>
 #include <tf/transform_datatypes.h>
+#include <pluginlib/class_loader.h>
+#include <meshing_plugins.h>
```
Absolute file path?

woodaaron replied:
Not clear on what you're saying here.

Jmeyer1292 commented on godel_surface_detection/src/detection/surface_detection.cpp:
```diff
@@ -317,13 +319,28 @@ namespace godel_surface_detection
+
+      try
+      {
+        mesher = poly_loader.createInstance("meshing_plugins::ConcaveHullMesher");
```
This may go without saying, but we'll need to make this a parameter.

Jmeyer1292 commented on godel_surface_detection/src/services/blending_service_path_generation.cpp:
```diff
@@ -384,122 +147,188 @@ SurfaceBlendingService::generateProcessPath(const int& id,
+{
+  std::vector<geometry_msgs::PoseArray> blend_result, edge_result, scan_result;
+
+  pluginlib::ClassLoader<path_planning_base::PathPlanningBase> poly_loader("path_planning_plugins",
```
Would you consider making the `ClassLoader` objects themselves a part of the SurfaceBlendingService?

woodaaron replied:
Sure. What's your reasoning behind this thought?

Jmeyer1292 commented on godel_surface_detection/src/services/blending_service_path_generation.cpp:
```diff
@@ -384,122 +147,188 @@ SurfaceBlendingService::generateProcessPath(const int& id,
+    nh.getParam(QUALITY_METRIC_PARAM, scan_params.quality_metric);
+    nh.getParam(WINDOW_WIDTH_PARAM, scan_params.window_width);
+    nh.getParam(MIN_QA_VALUE_PARAM, scan_params.min_qa_value);
+    nh.getParam(MAX_QA_VALUE_PARAM, scan_params.min_qa_value);
```
So we get a certain class of parameters from the request message, and the rest are loaded via parameters after each request? I imagine I'll find it in the GUI code, but are these updated when the user selects certain items in the menu?

Jmeyer1292 commented on meshing_plugins/include/meshing_base.h:
```diff
@@ -0,0 +1,34 @@
+#include <pcl/point_types.h>
+#include <pcl/PolygonMesh.h>
+
+namespace meshing_base
```
So from my limited knowledge, one typically creates a package that contains ONLY the headers related to the interface of the plugin. In a separate package one creates the implementation. The using package includes only a dependency on the definition package.

Jmeyer1292 commented on meshing_plugins/include/meshing_base.h:
```diff
@@ -0,0 +1,34 @@
+  public:
+    /**
+     * @brief Initialize the mesher
+     * @details [long description]
```
Please clean up the auto-generated documents.

Jmeyer1292 commented on meshing_plugins/include/meshing_plugins.h:
```diff
@@ -0,0 +1,27 @@
+#include <pcl/surface/ear_clipping.h>
+#include <geometry_msgs/PoseArray.h>
+
+const static double CONCAVE_HULL_ALPHA = 0.1;
```
Un-namespaced constant in a header. Could it live in the source?

Jmeyer1292 commented on meshing_plugins/include/meshing_plugins.h:
```diff
@@ -0,0 +1,27 @@
+  {
+  private:
+    pcl::ConcaveHull<meshing_base::Point> chull_;
+    pcl::EarClipping clipping_;
```
Have you considered just making the `chull_` objects and `clipping_` objects local variables inside the `generateMesh` function?

woodaaron replied:
Done.

Jmeyer1292 commented on path_planning_plugins/include/path_planning_base.h:
```diff
@@ -0,0 +1,19 @@
+
+namespace path_planning_base
+{
+  typedef  godel_msgs::PathPlanningParameters PlanningParams;
```
If it's not in the public interface, we might as well move it to the source file.

Jmeyer1292 commented on path_planning_plugins/include/path_planning_plugins.h:
```diff
@@ -0,0 +1,50 @@
+
+namespace Openveronoi
+{
+  bool populatePlanningParameters(ros::NodeHandle& nh, path_planning_base::PlanningParams& params);
```
Would a user ever call this?

woodaaron replied:
I don't anticipate a user would call it, but it's common to both ScanPlanner and BlendPlanner, and I wanted to avoid copy/paste. I could write it as a default implementation within path_planner_base and have both inherit it, or make it a private function in each subclass, but having it exist within the Openveronoi namespace seemed like a good compromise.

Jmeyer1292 commented on path_planning_plugins/src/path_planning_plugins.cpp:
```diff
@@ -0,0 +1,259 @@
+  void Openveronoi::BlendPlanner::init(pcl::PolygonMesh mesh)
+  {
+    mesh_ = mesh;
+    mesh_importer_ = new mesh_importer::MeshImporter(false);
```
For each `new` we should also `delete`, or better yet, use `std::unique_ptr`.

woodaaron replied:
On reflection, I'm not sure mesh_importer_ needs to be a private member. We really just use it to calculate boundaries and a pose. Once we get to the end of generatePath, there's no reason to hold on to it. I'll move it to generatePath and use a unique_ptr.

Commit:
Separated plugin implementations from base classes

Removed unused PCL dependency

Added orientation to pose on path planner

Moved concave hull and ear clipping to source and renamed references.

Cleaned up autogenerated code and included variable names in prototype

Made mesh_importer a unique pointer and changed from a private class member to a local object

Jmeyer1292 commented:
Thanks for your contribution.
